include/pybind11/functional.h

```diff
@@ -43,27 +43,37 @@ struct type_caster<std::function<Return(Args...)>> {
            captured variables), in which case the roundtrip can be avoided.
          */
         if (auto cfunc = func.cpp_function()) {
-            auto c = reinterpret_borrow<capsule>(PyCFunction_GET_SELF(cfunc.ptr()));
-            auto rec = (function_record *) c;
-
-            while (rec != nullptr) {
-                if (rec->is_stateless
-                    && same_type(typeid(function_type),
-                                 *reinterpret_cast<const std::type_info *>(rec->data[1]))) {
-                    struct capture {
-                        function_type f;
-                    };
-                    value = ((capture *) &rec->data)->f;
-                    return true;
+            auto cfunc_self = PyCFunction_GET_SELF(cfunc.ptr());
+            if (isinstance<capsule>(cfunc_self)) {
+                auto c = reinterpret_borrow<capsule>(cfunc_self);
+                auto rec = (function_record *) c;
+
+                while (rec != nullptr) {
+                    if (rec->is_stateless
+                        && same_type(typeid(function_type),
+                                     *reinterpret_cast<const std::type_info *>(rec->data[1]))) {
+                        struct capture {
+                            function_type f;
+                        };
+                        value = ((capture *) &rec->data)->f;
+                        return true;
+                    }
+                    rec = rec->next;
                 }
-                rec = rec->next;
+            } else {
+               // Usually indicates that it is a builtin function.
+#if            defined(PYPY_VERSION)
+               // PyPy will segfault otherwise when passing in raw builtin functions.
+               // This will lead to a TypeError instead.
+               return false;
```

Comment:
So it looks like this test case ALWAYS segfaulted on non-ubuntu PyPy then which means that that the code is still better than before. I am really stumped on how to avoid this segfault though.

Reply:
I just counted, we have 19 xfail for PyPy already.

Reply:
PyPy has bugs. If we aren't adding a new one or making it worse (which it appears we are not), then I'm happy with it if it passes Google's testing.